File: go/signers/cdp/signer.go
package cdp

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"net/http"
	"strconv"
	"unicode/utf8"

	"github.com/gagliardetto/solana-go"

	"github.com/solana-foundation/solana-keychain/go/core"
)

const (
	// defaultBaseURL is the production CDP API endpoint.
	defaultBaseURL = "https://api.cdp.coinbase.com"
	// basePath is the CDP Solana accounts base path.
	basePath = "/platform/v2/solana/accounts"
)

// Signer signs Solana transactions and messages with CDP's managed key
// infrastructure via the CDP REST API. The account address is provided at
// construction time; no remote calls happen until a signing method is invoked.
// All fields are immutable after New, so a Signer is safe for concurrent use.
type Signer struct {
	apiKeyID     string
	apiKeySecret string
	walletSecret string
	pubkey       solana.PublicKey
	apiBaseURL   string
	apiHost      string
	client       *http.Client
}

// Ensure Signer satisfies the core contract at compile time.
var _ core.Signer = (*Signer)(nil)

// New builds a CDP signer from cfg. Construction performs no I/O: key material
// is only validated when the first JWT is built.
func New(cfg Config) (*Signer, error) {
	if cfg.APIKeyID == "" {
		return nil, core.NewSignerError(core.CodeConfigError, "api_key_id must not be empty")
	}
	if cfg.APIKeySecret == "" {
		return nil, core.NewSignerError(core.CodeConfigError, "api_key_secret must not be empty")
	}
	if cfg.WalletSecret == "" {
		return nil, core.NewSignerError(core.CodeConfigError, "wallet_secret must not be empty")
	}
	if cfg.Address == "" {
		return nil, core.NewSignerError(core.CodeConfigError, "address must not be empty")
	}

	pubkey, err := solana.PublicKeyFromBase58(cfg.Address)
	if err != nil {
		return nil, core.WrapSignerError(core.CodeInvalidPublicKey, "invalid Solana address: "+cfg.Address, err)
	}

	baseURL, err := core.NormalizeHTTPSBaseURL(cfg.APIBaseURL, defaultBaseURL, "api_base_url")
	if err != nil {
		return nil, err
	}
	apiHost, err := core.HostFromURL(baseURL)
	if err != nil {
		return nil, err
	}

	client := cfg.HTTPClient
	if client == nil {
		client = core.NewHTTPClient(cfg.HTTPClientConfig)
	}

	return &Signer{
		apiKeyID:     cfg.APIKeyID,
		apiKeySecret: cfg.APIKeySecret,
		walletSecret: cfg.WalletSecret,
		pubkey:       pubkey,
		apiBaseURL:   baseURL,
		apiHost:      apiHost,
		client:       client,
	}, nil
}

// Pubkey returns the CDP-managed account's public key.
func (s *Signer) Pubkey() solana.PublicKey { return s.pubkey }

// String renders the signer without any secret material.
func (s Signer) String() string {
	return "cdp.Signer{pubkey: " + s.pubkey.String() + ", apiBaseURL: " + s.apiBaseURL + "}"
}

// GoString mirrors String so %#v cannot leak secrets either.
func (s Signer) GoString() string { return s.String() }

// SignMessage signs arbitrary bytes via the CDP signMessage endpoint and
// verifies the returned signature against this signer's public key.
//
// Quirk: the CDP signMessage API takes a UTF-8 string, so non-UTF-8 payloads
// are rejected with CodeSerializationError.
func (s *Signer) SignMessage(ctx context.Context, message []byte) (solana.Signature, error) {
	if !utf8.Valid(message) {
		return solana.Signature{}, core.NewSignerError(core.CodeSerializationError,
			"CDP signMessage requires UTF-8; non-UTF-8 bytes are not supported")
	}

	path := basePath + "/" + s.pubkey.String() + "/sign/message"
	var resp signMessageResponse
	if err := s.doPost(ctx, path, map[string]any{"message": string(message)}, &resp, "sign_message"); err != nil {
		return solana.Signature{}, err
	}

	sig, err := core.DecodeSignatureBase58(resp.Signature, "cdp")
	if err != nil {
		return solana.Signature{}, err
	}

	if err := core.VerifySignature(s.pubkey, message, sig); err != nil {
		return solana.Signature{}, err
	}
	return sig, nil
}

// SignTransaction sends the full serialized transaction (base64 wire format,
// with required-signature slots zero-filled) to the CDP signTransaction
// endpoint, then extracts only this signer's signature from the returned
// signed transaction, verifies it against the original message bytes, and
// applies it to tx.
func (s *Signer) SignTransaction(ctx context.Context, tx *solana.Transaction) (core.SignedTransaction, error) {
	msgBytes, err := tx.Message.MarshalBinary()
	if err != nil {
		return core.SignedTransaction{}, core.WrapSignerError(core.CodeSerializationError,
			"failed to serialize transaction message", err)
	}
	pos, err := core.SigningPosition(tx, s.pubkey)
	if err != nil {
		return core.SignedTransaction{}, err
	}

	// Serialize the full transaction (Solana wire format): the wire encoding
	// carries one slot per required signature, so pad a copy with zero-filled
	// slots.
	reqTx := *tx
	if n := int(tx.Message.Header.NumRequiredSignatures); len(reqTx.Signatures) < n {
		sigs := make([]solana.Signature, n)
		copy(sigs, tx.Signatures)
		reqTx.Signatures = sigs
	}
	serialized, err := reqTx.MarshalBinary()
	if err != nil {
		return core.SignedTransaction{}, core.WrapSignerError(core.CodeSerializationError,
			"failed to serialize transaction", err)
	}

	path := basePath + "/" + s.pubkey.String() + "/sign/transaction"
	body := map[string]any{"transaction": base64.StdEncoding.EncodeToString(serialized)}
	var resp signTransactionResponse
	if err := s.doPost(ctx, path, body, &resp, "sign_transaction"); err != nil {
		return core.SignedTransaction{}, err
	}

	// Decode and deserialize the returned signed transaction.
	signedBytes, err := base64.StdEncoding.DecodeString(resp.SignedTransaction)
	if err != nil {
		return core.SignedTransaction{}, core.WrapSignerError(core.CodeSerializationError,
			"failed to decode base64 signed transaction from CDP", err)
	}
	signedTx, err := solana.TransactionFromBytes(signedBytes)
	if err != nil {
		return core.SignedTransaction{}, core.WrapSignerError(core.CodeSerializationError,
			"failed to deserialize signed transaction from CDP", err)
	}

	// Extract only our signature from the response and apply it to the original
	// transaction.
	if pos >= len(signedTx.Signatures) {
		return core.SignedTransaction{}, core.NewSignerError(core.CodeSigningFailed,
			"signature not found at expected position in CDP response")
	}
	sig := signedTx.Signatures[pos]
	if err := core.VerifySignature(s.pubkey, msgBytes, sig); err != nil {
		return core.SignedTransaction{}, err
	}

	return core.AttachSignature(tx, s.pubkey, sig)
}

// IsAvailable checks that the CDP API is reachable and this account is
// accessible by fetching the account info (GET, bearer auth only). All errors
// are swallowed and reported as false.
func (s *Signer) IsAvailable(ctx context.Context) bool {
	ctx, cancel := context.WithTimeout(ctx, core.AvailabilityTimeout)
	defer cancel()
	path := basePath + "/" + s.pubkey.String()
	authToken, err := createAuthJWT(s.apiKeyID, s.apiKeySecret, s.apiHost, http.MethodGet, path)
	if err != nil {
		return false
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, s.apiBaseURL+path, nil)
	if err != nil {
		return false
	}
	req.Header.Set("Authorization", "Bearer "+authToken)

	status, _, err := core.SendRequest(s.client, req, "cdp")
	return err == nil && core.IsSuccess(status)
}

// doPost sends an authenticated POST to a CDP signing endpoint (bearer auth JWT
// plus X-Wallet-Auth wallet JWT) and decodes the 2xx JSON response into out.
// what names the endpoint in error details ("sign_message" / "sign_transaction").
func (s *Signer) doPost(ctx context.Context, path string, body map[string]any, out any, what string) error {
	bodyBytes, err := core.MarshalCanonicalJSON(body)
	if err != nil {
		return core.WrapSignerError(core.CodeSerializationError, "failed to serialize request body", err)
	}
	authToken, err := createAuthJWT(s.apiKeyID, s.apiKeySecret, s.apiHost, http.MethodPost, path)
	if err != nil {
		return err
	}
	walletToken, err := createWalletJWT(s.walletSecret, s.apiHost, http.MethodPost, path, body)
	if err != nil {
		return err
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, s.apiBaseURL+path, bytes.NewReader(bodyBytes))
	if err != nil {
		return core.WrapSignerError(core.CodeHTTPError, "failed to build CDP request", err)
	}
	req.Header.Set("Authorization", "Bearer "+authToken)
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-Wallet-Auth", walletToken)

	status, respBody, err := core.SendRequest(s.client, req, "cdp")
	if err != nil {
		return err
	}

	// Only the status code goes into the error; the response body is never
	// attached.
	if !core.IsSuccess(status) {
		return core.NewSignerError(core.CodeRemoteAPIError, "CDP API error "+strconv.Itoa(status))
	}

	if err := json.Unmarshal(respBody, out); err != nil {
		return core.WrapSignerError(core.CodeSerializationError, "failed to parse CDP "+what+" response", err)
	}
	return nil
}

File: python/src/solana_keychain/cdp/signer.py
"""CDP (Coinbase Developer Platform) signer integration."""

import base64
from dataclasses import dataclass, field
from typing import Any

try:
    import base58
except ImportError as error:  # pragma: no cover
    raise ImportError(
        "solana_keychain.cdp requires the cdp extra: pip install 'solana-keychain[cdp]'"
    ) from error

import httpx
from solders.pubkey import Pubkey
from solders.signature import Signature
from solders.transaction import VersionedTransaction

from solana_keychain.cdp.jwt import create_auth_jwt, create_wallet_jwt
from solana_keychain.core.errors import SignerError, SignerErrorCode
from solana_keychain.core.http import (
    assert_https_url,
    fetch_signer_json,
    normalize_base_url,
    probe_availability,
)
from solana_keychain.core.signature_util import (
    extract_and_verify_returned_signature,
    verify_returned_signature,
)
from solana_keychain.core.signer import SignedTransaction, SolanaSigner
from solana_keychain.core.transaction_util import (
    ED25519_SIGNATURE_LENGTH,
    add_signature_to_transaction,
    classify_signed_transaction,
    serialize_transaction,
    signed_message_bytes,
)
from solana_keychain.core.wallet_jwt import extract_host

DEFAULT_API_BASE_URL = "https://api.cdp.coinbase.com"
BASE_PATH = "/platform/v2/solana/accounts"


@dataclass
class CdpSignerConfig:
    """Configuration for a CDP signer.

    ``api_key_secret`` is the base64 Ed25519 API private key (seed ‖ pubkey);
    ``wallet_secret`` is the base64 PKCS8 DER P-256 wallet key used for the
    ``X-Wallet-Auth`` token on write endpoints; ``address`` is the base58 Solana
    account address managed by CDP.
    """

    api_key_id: str
    api_key_secret: str = field(repr=False)
    wallet_secret: str = field(repr=False)
    address: str
    api_base_url: str = DEFAULT_API_BASE_URL
    http_client: httpx.AsyncClient | None = field(default=None, repr=False)


class CdpSigner(SolanaSigner):
    """Signer backed by a CDP-managed Solana account.

    ``sign_message`` requires UTF-8 payloads — the sign-message endpoint takes a
    string, so arbitrary non-UTF-8 bytes are rejected.
    """

    def __init__(self, config: CdpSignerConfig) -> None:
        for name, value in (
            ("api_key_id", config.api_key_id),
            ("api_key_secret", config.api_key_secret),
            ("wallet_secret", config.wallet_secret),
            ("address", config.address),
        ):
            if not value:
                raise SignerError(SignerErrorCode.CONFIG_ERROR, f"{name} must not be empty")
        try:
            self._pubkey = Pubkey.from_string(config.address)
        except Exception:
            raise SignerError(
                SignerErrorCode.INVALID_PUBLIC_KEY, f"Invalid Solana address: {config.address}"
            ) from None
        api_base_url = normalize_base_url(config.api_base_url)
        assert_https_url(api_base_url, "api_base_url")
        self._api_base_url = api_base_url
        self._api_host = extract_host(api_base_url, "CDP")
        self._api_key_id = config.api_key_id
        self._api_key_secret = config.api_key_secret
        self._wallet_secret = config.wallet_secret
        self._http_client = config.http_client

    def __repr__(self) -> str:
        return f"CdpSigner(pubkey={self._pubkey}, api_base_url={self._api_base_url})"

    @property
    def pubkey(self) -> Pubkey:
        return self._pubkey

    async def _post_signed(self, path: str, body: dict[str, Any]) -> Any:
        auth_token = create_auth_jwt(
            self._api_key_id, self._api_key_secret, self._api_host, "POST", path
        )
        wallet_token = create_wallet_jwt(self._wallet_secret, self._api_host, "POST", path, body)
        return await fetch_signer_json(
            url=f"{self._api_base_url}{path}",
            provider_name="CDP",
            method="POST",
            headers={
                "Authorization": f"Bearer {auth_token}",
                "Content-Type": "application/json",
                "X-Wallet-Auth": wallet_token,
            },
            json_body=body,
            client=self._http_client,
        )

    async def sign_message(self, message: bytes) -> Signature:
        try:
            message_str = message.decode("utf-8")
        except UnicodeDecodeError:
            raise SignerError(
                SignerErrorCode.SERIALIZATION_ERROR,
                "CDP signMessage requires UTF-8; non-UTF-8 bytes are not supported",
            ) from None
        path = f"{BASE_PATH}/{self._pubkey}/sign/message"
        response = await self._post_signed(path, {"message": message_str})
        signature_b58 = response.get("signature") if isinstance(response, dict) else None
        if not isinstance(signature_b58, str):
            raise SignerError(
                SignerErrorCode.SERIALIZATION_ERROR, "Failed to parse CDP sign_message response"
            )
        try:
            signature_bytes = base58.b58decode(signature_b58)
        except ValueError:
            raise SignerError(
                SignerErrorCode.SERIALIZATION_ERROR,
                "Failed to decode base58 signature from CDP",
            ) from None
        if len(signature_bytes) != ED25519_SIGNATURE_LENGTH:
            raise SignerError(
                SignerErrorCode.SIGNING_FAILED,
                f"Invalid signature length from CDP (expected {ED25519_SIGNATURE_LENGTH} bytes)",
            )
        signature = Signature.from_bytes(signature_bytes)
        return verify_returned_signature(signature, self._pubkey, message)

    async def sign_transaction(self, transaction: VersionedTransaction) -> SignedTransaction:
        message_data = signed_message_bytes(transaction.message)
        path = f"{BASE_PATH}/{self._pubkey}/sign/transaction"
        encoded_tx = base64.b64encode(bytes(transaction)).decode("ascii")
        response = await self._post_signed(path, {"transaction": encoded_tx})

        signed_b64 = response.get("signedTransaction") if isinstance(response, dict) else None
        if not isinstance(signed_b64, str):
            raise SignerError(
                SignerErrorCode.SERIALIZATION_ERROR,
                "Failed to parse CDP sign_transaction response",
            )
        try:
            signed_bytes = base64.b64decode(signed_b64, validate=True)
        except Exception:
            raise SignerError(
                SignerErrorCode.SERIALIZATION_ERROR,
                "Failed to decode base64 signed transaction from CDP",
            ) from None
        signature = extract_and_verify_returned_signature(
            signed_bytes, self._pubkey, message_data, "CDP"
        )
        add_signature_to_transaction(transaction, self._pubkey, signature)
        return classify_signed_transaction(
            transaction, serialize_transaction(transaction), signature
        )

    async def is_available(self) -> bool:
        path = f"{BASE_PATH}/{self._pubkey}"

        async def probe() -> bool:
            auth_token = create_auth_jwt(
                self._api_key_id, self._api_key_secret, self._api_host, "GET", path
            )
            await fetch_signer_json(
                url=f"{self._api_base_url}{path}",
                provider_name="CDP",
                headers={"Authorization": f"Bearer {auth_token}"},
                client=self._http_client,
            )
            return True

        return await probe_availability(probe)


async def create_cdp_signer(config: CdpSignerConfig) -> CdpSigner:
    """Create a ready-to-use CDP signer."""
    return CdpSigner(config)
